src/Build/BackEnd/Components/SdkResolution/NuGet/NuGetSdkResolver.cs

```diff
@@ -0,0 +1,202 @@
+            context.State = msbuildSdkVersions;
+
+            string globalJsonVersion;
+
+            // Check if global.json specified a version for this SDK and make sure its a version compatible with NuGet
+            if (msbuildSdkVersions != null && msbuildSdkVersions.TryGetValue(id, out globalJsonVersion) && !String.IsNullOrWhiteSpace(globalJsonVersion))
+            {
+                return NuGetAbstraction.TryParseNuGetVersion(globalJsonVersion, out parsedVersion);
+            }
+
+            parsedVersion = null;
+
+            return false;
+        }
+
+        /// <summary>
+        /// IMPORTANT: This class is used to ensure that <see cref="NuGetSdkResolver"/> does not consume any NuGet classes directly.  This ensures that no NuGet assemblies
+        /// are loaded unless they are needed.  Do not implement anything in <see cref="NuGetSdkResolver"/> that uses a NuGet class and instead place it here.
+        /// </summary>
+        private static class NuGetAbstraction
+        {
+            public static SdkResultBase GetSdkResult(SdkReference sdk, object nuGetVersion, SdkResolverContextBase context, SdkResultFactoryBase factory)
+            {
+                // Cast the NuGet version since the caller does not want to consume NuGet classes directly
+                NuGetVersion parsedSdkVersion = (NuGetVersion) nuGetVersion;
+
+                // Stores errors and warnings for the result
+                ICollection<string> errors = new List<string>();
+                ICollection<string> warnings = new List<string>();
+
+                // Load NuGet settings and a path resolver
+                ISettings settings = Settings.LoadDefaultSettings(context.ProjectFilePath);
+
+                FallbackPackagePathResolver fallbackPackagePathResolver = new FallbackPackagePathResolver(NuGetPathContext.Create(settings));
+
+                string installedPath, installedVersion;
+
+                // Attempt to find a package if its already installed
+                if (!TryGetMSBuildSdkPackageInfo(fallbackPackagePathResolver, sdk.Name, parsedSdkVersion, out installedPath, out installedVersion))
+                {
+                    try
+                    {
+                        context.Logger.LogMessage($"Resolving SDK '{sdk}'...");
+
+                        // Asynchronously run the restore without a commit which find the package on configured feeds, download, and unzip it without generating any other files
+                        IReadOnlyList<RestoreResultPair> results = RestoreRunnerEx.RunWithoutCommit(
+                                context.ProjectFilePath,
+                                sdk.Name,
+                                parsedSdkVersion.ToFullString(),
+                                settings,
+                                new NuGetSdkLogger(context.Logger, warnings, errors))
+                            .Result;
```

Comment:
Does this need a `ConfigureAwait`? (I'm really asking; I don't grok when that is needed.)